### PR TITLE
Restore MySQL driver and fix ordering problem

### LINKED ISSUE #856: Ordering behavior is incorrect/inconsistent on some database drivers
`orderby` appears to produce inconsistent results (as already documented in `factorials.links`).  A minimal example:
```
var db = database "links";
var factorials = table "factorials" with (i : Int, f : Int) from db;

fun insertSome() {
  delete (x <-- factorials);
  insert factorials
  values (f, i)
     [(f=1, i=1),(f=2, i=2),(f=6, i=3)];
}
var asdf = insertSome();
query {
    for (row <-- factorials)
    where (row.i <= 5)
    orderby (row.i)
      [(i=row.i, f=row.f)]
  }
```
We get the result list in reverse order in Postgres and MySQL, while for SQLite3 the order is correct.

The SQL query that is generated for each database is the same, so presumably somewhere in the Postgres and MySQL database drivers the order of the returned results is being reversed.

This blocks #322 since it prevents us running the tests on all three databases and getting the same results.

### PR DESCRIPTION
**Pull request by jamescheney**
This PR restores the MySQL driver, a precondition for #322, and fixes #856 (albeit this could be cleaned up a bit).

Marked as WIP because the way query results are constructed by Sqlite3 is still a little inefficient and could be improved, and I'd like to make progress towards #855 also before merging, unless it turns out to be harder than I think.

- [x] Avoid unnecessary `List.rev`s in sqlite3 driver and value.ml
- [X] Basic support for INSERT RETURNING for all supported dbs.
- [x] Delete remaining dead code

### CODE REVIEW

**Comment by jamescheney** [comment]
This fixes parts of #856 except for the extension to support RETURNING multiple columns which I leave as a todo.

Something horrible has happened so the PR / branch probably needs to be redone from a clean checkout of master.

**Comment by SimonJF** [comment]
Regarding something horrible happening, have you tried:

> Here's a good workaround. Use the Edit button when viewing the PR in GitHub to change the base branch to something other than master. Then switch it back to master and it will now correctly show only the changes from the most recent commits.

(from the following link; there are also some other solutions)
 
https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch

**Comment by jamescheney** [comment]
I seem to have found / remembered the incantation needed to get the branches synced up again, which was `git merge --no-ff master` - if that doesn't cause problems I'm inclined to leave as is.

**Comment by SimonJF** [comment]
Ah, good to know!

**Comment by jamescheney** [comment]
So the CI is failing because of an apparent regression on #835...

**Comment by jamescheney** [comment]
I take that back, I only thought that because I had a stale build.  Looks like I just needed to adjust the factorials example again.

**Comment by jamescheney** [comment]
I've decided to weaken #855 sufficiently that this fixes it.

**Comment by jamescheney** [comment]
Now fixed sqlite3 inefficiency and realized that mysql driver suffered from a different problem (whose solution is similar).  Previously, we would randomly access the mysql library's result which was apparently slow.  Now, instead I load it all into an append-only buffer.  Possibly doing this for postgres results would be better and more uniform but this PR is already getting too big.

**Comment by jamescheney** [comment]
The code for handling null integers in mySQL was broken, this is now fixed and I've added a database test that checks that null handling works correctly, as a first step towards #655.  This should be ready for review now.

File: core/database.ml [hunk 1]
```diff
@@ -91,13 +90,10 @@ let execute_insert_returning returning q db =
...
               match result#status with
                | `QueryOk ->
-                   let rows = result#get_all_lst in
-                     begin
-                       match rows with
-                         | [[id]] -> Value.box_int (int_of_string id)
-                         | _ ->
-                             raise (runtime_error ("Returned the wrong number of results executing " ^ q))
-                     end
+                  if result#nfields == 1 && result#ntuples == 1
```
Comment (dhil) [inline core/database.ml:93]: Structural equality `=`.

File: core/utility.ml [hunk 2]
```diff
@@ -1458,3 +1458,66 @@ struct
...
+    then Array.get (Array.get buf.pages (i/buf.pagesize)) (i mod buf.pagesize)
+    else raise Not_found
+
+  let append buf x =
+    (* first, check if there is enough space or allocation is needed *)
+    if (buf.nextitem == buf.pagesize)
+    then begin
+      buf.nextitem <- 0;
+      buf.currpage <- buf.currpage+1;
+      if (buf.currpage == buf.numpages)
```
Comment (dhil) [inline core/utility.ml:1509]: Structural equality `=`.

File: core/utility.ml [hunk 3]
```diff
@@ -1458,3 +1458,66 @@ struct
...
+                    currpage = 0;
+                    nextitem = 0;
+                    pages = Array.init n (fun _ -> Array.init m (fun _ -> x)) }
+
+  let length buf = buf.currpage*buf.pagesize + buf.nextitem
+
+  let set buf i x =
+  if 0 <= i && i < buf.currpage*buf.pagesize + buf.nextitem
+    then Array.set (Array.get buf.pages (i/buf.pagesize)) (i mod buf.pagesize) x
+    else raise Not_found
```
Comment (dhil) [inline core/utility.ml:1496]: Maybe use `raise (Invalid_argument "out of bounds")`.

File: core/utility.ml [hunk 4]
```diff
@@ -1458,3 +1458,66 @@ struct
...
+    then Array.get (Array.get buf.pages (i/buf.pagesize)) (i mod buf.pagesize)
+    else raise Not_found
+
+  let append buf x = 
+    (* first, check if there is enough space or allocation is needed *)
+    if (buf.nextitem == buf.pagesize)
+    then begin 
+      buf.nextitem <- 0;
+      buf.currpage <- buf.currpage+1;
+      if (buf.currpage == buf.numpages)
```
[the remark this answers is not in the document]
Comment (jamescheney) [reply core/utility.ml]: Done.  (But structural equality on integers behaves the same as `==`.)

File: core/utility.ml (hunk 3 above)
Comment (jamescheney) [reply core/utility.ml:1496]: OK, matching what OCaml does with ordinary arrays.

**Comment by jamescheney** [comment]
This should await #859.

**Comment by jamescheney** [comment]
Miraculously, this seems to have commited with 859 and #861, so now that they are merged, will merge this once CI passes if no one stops me.

**Pull request merged**